lxd/device/gpu_physical.go

```diff
@@ -342,16 +527,82 @@ func (d *gpuPhysical) pciDeviceDriverOverrideIOMMU(pciDev pcidev.Device, driverO
 	return nil
 }
 
+func (d *gpuPhysical) stopCDIDevices(configDevices *cdi.ConfigDevices, runConf *deviceConfig.RunConfig) error {
+	for deviceName := range configDevices.UnixCharDevs {
+		err := unixDeviceRemove(d.inst.DevicesPath(), "unix", deviceName, "", runConf)
+		if err != nil {
+			return err
+		}
+
+		runConf.PostHooks = append(runConf.PostHooks, func() error {
+			err := unixDeviceDeleteFiles(d.state, d.inst.DevicesPath(), "unix", deviceName, "")
+			if err != nil {
+				return fmt.Errorf("Failed to delete files for device '%s': %w", deviceName, err)
+			}
+
+			return nil
+		})
+	}
+
+	for deviceName, conf := range configDevices.DiskDevs {
+		relativeDestPath := strings.TrimPrefix(conf["path"], "/")
+		devPath := filepath.Join(d.inst.DevicesPath(), filesystem.PathNameEncode(deviceJoinPath("disk", deviceName, relativeDestPath)))
+		runConf.PostHooks = append(runConf.PostHooks, func() error {
+			// Clean any existing device mount entry. Should occur first before custom volume unmounts.
+			err := DiskMountClear(devPath)
+			if err != nil {
+				return err
+			}
+
+			return nil
+		})
+
+		// The disk device doesn't exist do nothing.
+		if !shared.PathExists(devPath) {
+			return nil
+		}
+
+		// Request an unmount of the device inside the instance.
+		runConf.Mounts = append(runConf.Mounts, deviceConfig.MountEntryItem{
+			TargetPath: relativeDestPath,
+		})
+	}
+
+	return nil
+}
+
 // Stop is run when the device is removed from the instance.
 func (d *gpuPhysical) Stop() (*deviceConfig.RunConfig, error) {
 	runConf := deviceConfig.RunConfig{
 		PostHooks: []func() error{d.postStop},
 	}
 
 	if d.inst.Type() == instancetype.Container {
-		err := unixDeviceRemove(d.inst.DevicesPath(), "unix", d.name, "", &runConf)
-		if err != nil {
-			return nil, err
+		v := d.volatileGet()
+		if shared.IsTrue(v["cdi"]) {
+			cdiID, err := cdi.ToCDI(d.config["id"])
+			if cdiID.Empty() {
+				if err == nil {
+					err = fmt.Errorf(`Passed GPU device "id" config option is not compliant with the CDI notation`)
```

Comment:
Lets just return this error rather than appending to the "could not generate..." error below

Reply:
fixed